optional_plugins/varianter_cit/avocado_varianter_cit/__init__.py

```diff
@@ -0,0 +1,331 @@
+# Authors: Amador Pahim <amador@pahim.org>
+#          Bestoun S. Ahmed <bestoon82@gmail.com>
+#          Cleber Rosa <crosa@redhat.com>
+
+import copy
+import hashlib
+import itertools
+import os
+import random
+import sys
+
+from six import iteritems
+from six.moves import configparser
+
+from avocado.core import exit_codes
+from avocado.core.output import LOG_UI
+from avocado.core.plugin_interfaces import CLI
+from avocado.core.plugin_interfaces import Varianter
+from avocado.core.tree import TreeNode
+
+
+class VarianterCitCLI(CLI):
+
+    """
+    CIT Varianter options
+    """
+
+    name = 'cit'
+    description = "CIT Varianter options for the 'run' subcommand"
+
+    def configure(self, parser):
+
+        for name in ("run", "variants"):
+            subparser = parser.subcommands.choices.get(name, None)
+            if subparser is None:
+                continue
+            cit = subparser.add_argument_group('CIT varianter options')
+            cit.add_argument('--cit-parameter-file', metavar='PATH',
+                             help="Paths to a parameter file")
+            cit.add_argument('--cit-parameter-path', metavar='PATH',
+                             default='/run',
+                             help=('Default path for parameters generated '
+                                   'on the CIT variants'))
+            cit.add_argument('--cit-order-of-combinations',
+                             metavar='ORDER', type=int, default=2,
+                             help=("Order of combinations. Defaults to "
+                                   "%(default)s, maximum number is specific "
+                                   "to parameter file content"))
+
+    def run(self, args):
+        pass
+
+
+class VarianterCit(Varianter):
+
+    """
+    Processes the parameters file into variants
+    """
+
+    name = 'cit'
+    description = "CIT Varianter"
+
+    def initialize(self, args):
+        self.variants = None
+
+        cit_parameter_file = getattr(args, "cit_parameter_file", None)
+        if cit_parameter_file is None:
+            return
+        else:
+            cit_parameter_file = os.path.expanduser(cit_parameter_file)
+            if not os.access(cit_parameter_file, os.R_OK):
+                LOG_UI.error("parameter file '%s' could not be found or "
+                             "is not readable", cit_parameter_file)
+                self.error_exit(args)
+
+        self.parameter_path = getattr(args, "cit_parameter_path")
+        config = configparser.ConfigParser()
+        try:
+            config.read(cit_parameter_file)
+        except Exception as details:
+            LOG_UI.error("Cannot parse parameter file: %s", details)
+            self.error_exit(args)
+
+        parameters = [(key, value.split(', '))
+                      for key, value in config.items('parameters')]
+        constraints = [(key, value.split(', '))
+                       for key, value in config.items('constraints')]
```

Comment:
Perhaps it'd be useful to init empty one if `constrains` is not defined, especially when they are not supported (yet)

Reply:
I think that for the sake of readability (my own, hopefully for others), this could become:

```diff
-        parameters = [(key, value.split(', '))
-                      for key, value in config.items('parameters')]
-        constraints = [(key, value.split(', '))
-                       for key, value in config.items('constraints')]
+        parameters = self._get_key_values_from_config(config, 'parameters')
+        constraints = self._get_key_values_from_config(config, 'constraints')
         order = args.cit_order_of_combinations
         cit = Cit(parameters, constraints, order)
         self.headers, self.variants = cit.combine()
 
+    @staticmethod
+    def _get_key_values_from_config(config, section):
+        """
+        Returns the key and values from a config file section
+        """
+        return [(key, value.split(', '))
+                for key, value in config.items(section)]
+        

```

Reply:
BTW, @ldoktor I failed to understand the part about "to init empty one".  Do you mean set the list as None?